hazelcast/src/test/java/com/hazelcast/internal/ascii/MemcachedRawTcpTest.java

```diff
@@ -0,0 +1,614 @@
+                receiveLine(); // foo
+                receiveLine(); // END
+                fail();
+            }
+        });
+    }
+
+    @Test
+    public void testSetGet_withLargeValue() throws Exception {
+        int capacity = 10000;
+        StringBuilder value = new StringBuilder(capacity);
+        while (value.length() < capacity) {
+            value.append(randomString());
+        }
+
+        sendLine("set key 0 0 " + value.length());
+        sendLine(value.toString());
+        assertEquals("STORED", receiveLine());
+
+        sendLine("get key");
+        assertEquals("VALUE key 0 " + value.length(), receiveLine());
+        assertEquals(value.toString(), receiveLine());
+    }
+
+    @Test
+    public void testBulkSetGet_withManyKeys() throws Exception {
+        int numberOfKeys = 1000;
+
+        Map<String, String> entries = new HashMap<>();
+        for (int i = 0; i < numberOfKeys; i++) {
+            String key = "key" + i;
+            String value = "value" + i;
+            entries.put(key, value);
+
+            sendLine(String.format("set %s 0 0 %d", key, value.length()));
+            sendLine(value);
+            assertEquals("STORED", receiveLine());
+        }
+
+        sendLine("get " + String.join(" ", entries.keySet()));
+
+        for (int i = 0; i < numberOfKeys; i++) {
+            Matcher matcher = valueHeaderPattern.matcher(receiveLine());
+            assertTrue(matcher.matches());
+
+            String key = matcher.group(1);
+            int flags = Integer.parseInt(matcher.group(2));
+            int length = Integer.parseInt(matcher.group(3));
+
+            assertEquals(0, flags);
+
+            String value = receiveLine();
+            assertEquals(length, value.length());
+            assertEquals(entries.get(key), value);
+
+            // remove the entry to not count it again
+            entries.remove(key);
+        }
+        assertEquals("END", receiveLine());
+    }
+
+    /**
+     * Sends a single request line to the server.
+     * Adds CRLF at the end and flushes.
+     * @param line line to send
+     */
+    private void sendLine(String line) {
+        writer.write(line + CRLF);
+        writer.flush();
+    }
+
+    /**
+     * Tries to receive a response line from the server,
+     * calls Assert.fail() if nothing was received in 1 second.
+     * @return received line
+     */
+    private String receiveLine() throws Exception {
+        CompletableFuture<String> lineFuture = new CompletableFuture<>();
+
+        Thread t = new Thread(() -> {
+            if (scanner.hasNextLine()) {
+                lineFuture.complete(scanner.nextLine());
+            } else {
+                lineFuture.complete(null);
+            }
+        });
+        t.start();
+
+        String line = null;
+        try {
+            line = lineFuture.get(1000, TimeUnit.MILLISECONDS);
+        } catch (TimeoutException e) {
+            t.interrupt();
```

Comment:
I am not sure if the `hasNextLine` is interruptable but I assume the thread we have created will be ended after the input stream of the socket is closed in the tearDown

Reply:
How do we know whether sth is interruptable in Java? The documentation says it can block, so I assume it does some sort of I/O call at some point (poll, select etc.).
If you can think of a better way to implement this method, I can do that. This was the best I could come up with.

Reply:
I don't know a way either. I just started a `nc` server that just listens and wrote a similar code to yours and saw that `hasNextLine` is not interrupted. It returns when in fact there is a next line or the input stream is closed, which happens in the tearDown method anyway.

Reply:
Thanks for checking. Anyways, I removed the timeout logic, it will simply block forever if no there is no data to read.